bio-submission.txt

```diff
@@ -179,6 +179,12 @@ endio handler. The original value in this case was `end_workqueue_bio`. That
 function in turns is responsible to queueing the final execution of the real
 endio handler to the appropriate workqueue.
 
+Please note that `btrfs_end_bio` is set to a bio struct which represents a
+single stripe after the split made during the bio mapping. Therefore it's
+called separately for each stripe, despite the fact that it gets the full bio
+struct as an argument, which contains information about all stripes. It's
+executed in the softirq context.
+
```

Comment:
Instead of having this paragraph, how about the following diff: 

```
diff --git a/bio-submission.txt b/bio-submission.txt
index a0b25dc806ec..5cf9ed5b9212 100644
--- a/bio-submission.txt
+++ b/bio-submission.txt
@@ -164,19 +164,19 @@ saving the original end io handler and original private context of the bio into
 another structure - `btrfs_end_io_wq` and setting it as the `bi_private`
 information for the bio.
 
-Subsequently the bios is sent to be mapped as described above and right
+Subsequently the bio is sent to be mapped as described above and right
 before submission, in `submit_stripe_bio` the `bi_end_io`/`bi_private` are
 again changed for another set of data. Namely, the endio function is now set to
 `btrfs_end_bio` and bi_private is set to `struct btrfs_bio`. Given this, the
-first function which is run o bio completion is `btrfs_end_bio` it performs
-some generic bookeping, ensuring that if we have a request which involves multiple
-bios (i.e a write to RAID1) only the final bio would run further code. It also deals
-with incrementing error counters if the bio has encountered an error. If all is
-well eventually `btrfs_end_bbio` is called. It simply copies the original values of
-`bi_private`/`bi_end_io` to the bio and calls `bio_endio` which calls the new
-endio handler. The original value in this case was `end_workqueue_bio`. That
-function in turns is responsible to queueing the final execution of the real
-endio handler to the appropriate workqueue.
+first function which is run o bio completion is `btrfs_end_bio`. Since it's called
+for every stripe bio it first performs some generic bookeping, ensuring that
+for a request involving multiple bios (i.e a write to RAID1) only the final bio
+would run further code. It also deals with incrementing error counters if the
+bio has encountered an error. If all is well eventually `btrfs_end_bbio` is
+called. It simply copies the original values of `bi_private`/`bi_end_io` to the
+bio and calls `bio_endio` which calls the new endio handler. The original value
+in this case was `end_workqueue_bio`. That function in turns is responsible to
+queueing the final execution of the real endio handler to the appropriate workqueue.

```
It explicitly states that btrfs_end_bio is called once for every stripe bio. The reason why I would prefer it that way is that it's more logically connected to surrounding text.